File: sample-videochat-conference-java/src/main/java/com/quickblox/sample/videochat/conference/java/fragments/AudioConversationFragment.java
package com.quickblox.sample.videochat.conference.java.fragments;

import android.util.Log;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.widget.TextView;
import android.widget.ToggleButton;

import com.quickblox.conference.ConferenceSession;
import com.quickblox.sample.videochat.conference.java.R;
import com.quickblox.sample.videochat.conference.java.activities.CallActivity;
import com.quickblox.sample.videochat.conference.java.adapters.OpponentsFromCallAdapter;
import com.quickblox.videochat.webrtc.QBRTCAudioTrack;
import com.quickblox.videochat.webrtc.callbacks.QBRTCClientAudioTracksCallback;

import java.io.Serializable;


public class AudioConversationFragment extends BaseConversationFragment implements Serializable, OpponentsFromCallAdapter.OnAdapterEventListener,
        QBRTCClientAudioTracksCallback<ConferenceSession>, CallActivity.OnChangeDynamicToggle {
    private String TAG = getClass().getSimpleName();

    private TextView localName;
    private ToggleButton audioSwitchToggleButton;

    private boolean headsetPlugged;

    @Override
    public void onStart() {
        super.onStart();
        conversationFragmentCallbackListener.addOnChangeDynamicToggle(this);
    }

    @Override
    protected void initViews(View view) {
        localName = (TextView) view.findViewById(R.id.localName);
        localName.setVisibility(View.VISIBLE);
        audioSwitchToggleButton = (ToggleButton) view.findViewById(R.id.toggle_speaker);
        audioSwitchToggleButton.setVisibility(View.VISIBLE);
        super.initViews(view);
    }

    @Override
    public void onLocalAudioTrackReceive(ConferenceSession session, QBRTCAudioTrack audioTrack) {
        Log.d(TAG, "onLocalAudioTrackReceive() run");
        setStatusForCurrentUser(getString(R.string.text_status_connected));
        actionButtonsEnabled(true);
    }

    @Override
    public void onRemoteAudioTrackReceive(ConferenceSession session, QBRTCAudioTrack audioTrack, final Integer userID) {
        Log.d(TAG, "onRemoteAudioTrackReceive() run");
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        super.onCreateOptionsMenu(menu, inflater);
        MenuItem cameraSwitchItem = menu.findItem(R.id.camera_switch);
        cameraSwitchItem.setVisible(false);
    }

    @Override
    protected void initButtonsListener() {
        super.initButtonsListener();

        audioSwitchToggleButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                conversationFragmentCallbackListener.onSwitchAudio();
            }
        });
    }

    @Override
    protected void actionButtonsEnabled(boolean inability) {
        super.actionButtonsEnabled(inability);
        if (!headsetPlugged) {
            audioSwitchToggleButton.setEnabled(inability);
        }
        audioSwitchToggleButton.setActivated(inability);
    }

    @Override
    public void enableDynamicToggle(boolean plugged, boolean previousDeviceEarPiece) {
        headsetPlugged = plugged;
        audioSwitchToggleButton.setEnabled(!plugged);

        if (plugged) {
            audioSwitchToggleButton.setChecked(true);
        } else if (previousDeviceEarPiece) {
            audioSwitchToggleButton.setChecked(true);
        } else {
            audioSwitchToggleButton.setChecked(false);
        }
    }

    @Override
    protected void initTrackListeners() {
        super.initTrackListeners();
        initAudioTracksListener();
    }

    @Override
    protected void removeTrackListeners() {
        removeAudioTracksListener();
    }

    private void initAudioTracksListener() {
        if (currentSession != null) {
            currentSession.addAudioTrackCallbacksListener(this);
        }
    }

    private void removeAudioTracksListener() {
        if (currentSession != null) {
            currentSession.removeAudioTrackCallbacksListener(this);
        }
    }
}

File: sample-videochat-conference-kotlin/src/main/java/com/quickblox/sample/videochat/conference/kotlin/fragments/AudioConversationFragment.kt
package com.quickblox.sample.videochat.conference.kotlin.fragments

import android.util.Log
import android.view.Menu
import android.view.MenuInflater
import android.view.View
import android.widget.TextView
import android.widget.ToggleButton
import com.quickblox.conference.ConferenceSession
import com.quickblox.sample.videochat.conference.kotlin.R
import com.quickblox.sample.videochat.conference.kotlin.activities.CallActivity
import com.quickblox.sample.videochat.conference.kotlin.adapter.OpponentsFromCallAdapter
import com.quickblox.videochat.webrtc.QBRTCAudioTrack
import com.quickblox.videochat.webrtc.callbacks.QBRTCClientAudioTracksCallback
import java.io.Serializable


class AudioConversationFragment : BaseConversationFragment(), Serializable,
        OpponentsFromCallAdapter.OnAdapterEventListener,
        QBRTCClientAudioTracksCallback<ConferenceSession>, CallActivity.OnChangeDynamicToggle {
    private val TAG = AudioConversationFragment::class.simpleName

    private lateinit var localName: TextView
    private lateinit var audioSwitchToggleButton: ToggleButton

    private var headsetPlugged: Boolean = false

    override fun onStart() {
        super.onStart()
        conversationFragmentCallbackListener?.addOnChangeDynamicToggle(this)
    }

    override fun initViews(view: View) {
        localName = view.findViewById<View>(R.id.localName) as TextView
        localName.visibility = View.VISIBLE
        audioSwitchToggleButton = view.findViewById<View>(R.id.toggle_speaker) as ToggleButton
        audioSwitchToggleButton.visibility = View.VISIBLE
        super.initViews(view)
    }

    override fun onLocalAudioTrackReceive(session: ConferenceSession, audioTrack: QBRTCAudioTrack) {
        Log.d(TAG, "onLocalAudioTrackReceive() run")
        setStatusForCurrentUser(getString(R.string.text_status_connected))
        actionButtonsEnabled(true)
    }

    override fun onRemoteAudioTrackReceive(session: ConferenceSession, audioTrack: QBRTCAudioTrack, userID: Int?) {
        Log.d(TAG, "onRemoteAudioTrackReceive() run")
    }

    override fun onCreateOptionsMenu(menu: Menu?, inflater: MenuInflater?) {
        super.onCreateOptionsMenu(menu, inflater)
        val cameraSwitchItem = menu?.findItem(R.id.camera_switch)
        cameraSwitchItem?.isVisible = false
    }

    override fun initButtonsListener() {
        super.initButtonsListener()
        audioSwitchToggleButton.setOnClickListener { conversationFragmentCallbackListener?.onSwitchAudio() }
    }

    override fun actionButtonsEnabled(inability: Boolean) {
        super.actionButtonsEnabled(inability)
        if (!headsetPlugged) {
            audioSwitchToggleButton.isEnabled = inability
        }
        audioSwitchToggleButton.isActivated = inability
    }

    override fun enableDynamicToggle(plugged: Boolean, wasEarpiece: Boolean) {
        headsetPlugged = plugged
        audioSwitchToggleButton.isEnabled = !plugged
        if (plugged) {
            audioSwitchToggleButton.isChecked = true
        } else audioSwitchToggleButton.isChecked = wasEarpiece
    }

    override fun initTrackListeners() {
        super.initTrackListeners()
        initAudioTracksListener()
    }

    override fun removeTrackListeners() {
        removeAudioTracksListener()
    }

    private fun initAudioTracksListener() {
        if (currentSession != null) {
            currentSession?.addAudioTrackCallbacksListener(this)
        }
    }

    private fun removeAudioTracksListener() {
        if (currentSession != null) {
            currentSession?.removeAudioTrackCallbacksListener(this)
        }
    }
}